### ISSUE TITLE
[Bug]: Chunked embeddings break dot-product scoring with `use_activation=true

### ISSUE THREAD

**Issue by Ningke-Li** [vllm-project/vllm#57927]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Collecting environment information...
==============================
        System Info
==============================
OS                           : Ubuntu 22.04.5 LTS (x86_64)
GCC version                  : (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version                : Could not collect
CMake version                : version 3.22.1
Libc version                 : glibc-2.35

==============================
       PyTorch Info
==============================
PyTorch version              : 2.13.0+cu130
Is debug build               : False
CUDA used to build PyTorch   : 13.0
ROCM used to build PyTorch   : N/A
XPU used to build PyTorch    : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.14 | packaged by Anaconda, Inc. | (main, Aug 27 2026, 14:46:43) [GCC 14.3.0] (64-bit runtime)
Python platform              : Linux-5.15.0-122-generic-x86_64-with-glibc2.35
    
==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : Could not collect
CUDA_MODULE_LOADING set to   : 
GPU models and configuration : 
GPU 0: NVIDIA L40S
GPU 1: NVIDIA L40S
GPU 2: NVIDIA L40S
GPU 3: NVIDIA L40S

Nvidia driver version        : 595.71.05
cuDNN version                : Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : False

==============================
          CPU Info
==============================
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        52 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               128
On-line CPU(s) list:                  0-127
Vendor ID:                            AuthenticAMD
Model name:                           AMD EPYC 9554 64-Core Processor
CPU family:                           25
Model:                                17
Thread(s) per core:                   2
Core(s) per socket:                   64
Socket(s):                            1
Stepping:                             1
Frequency boost:                      enabled
CPU max MHz:                          3762.9880
CPU min MHz:                          1500.0000
BogoMIPS:                             6190.64
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d
Virtualization:                       AMD-V
L1d cache:                            2 MiB (64 instances)
L1i cache:                            2 MiB (64 instances)
L2 cache:                             64 MiB (64 instances)
L3 cache:                             256 MiB (8 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-127
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.6.16.post3
[pip3] nccl4py==0.5.0
[pip3] numpy==2.3.5
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cccl==13.3.3.4.1
[pip3] nvidia-cuda-crt==13.3.73
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvcc==13.3.73
[pip3] nvidia-cuda-nvdisasm==13.3.73
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cudnn-frontend==1.28.0
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufile==1.15.1.6
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-cutlass-dsl==4.6.2
[pip3] nvidia-cutlass-dsl-libs-base==4.6.2
[pip3] nvidia-cutlass-dsl-libs-core==4.6.2
[pip3] nvidia-cutlass-dsl-libs-cu12==4.6.2
[pip3] nvidia-cutlass-dsl-libs-cu13==4.6.2
[pip3] nvidia-ml-py==13.610.43
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.3.33
[pip3] nvidia-nvshmem-cu13==3.4.5
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvvm==13.3.73
[pip3] pyzmq==27.2.0
[pip3] tokenspeed-triton==3.8.10.post20260906
[pip3] torch==2.13.0
[pip3] torch_c_dlpack_ext==0.1.5
[pip3] torchaudio==2.11.0
[pip3] torchcodec==0.16.0
[pip3] torchvision==0.28.0
[pip3] transformers==5.16.1
[pip3] triton==3.7.1
[conda] numpy                        2.4.3            pypi_0              pypi

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.29.0
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled; XPU: Disabled
GPU Topology:
  	GPU0	GPU1	GPU2	GPU3	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	PHB	NODE	NODE	0-127	0		N/A
GPU1	PHB	 X 	NODE	NODE	0-127	0		N/A
GPU2	NODE	NODE	 X 	PHB	0-127	0		N/A
GPU3	NODE	NODE	PHB	 X 	0-127	0		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

==============================
     Environment Variables
==============================
NO_COLOR=1
CUDA_HOME=/usr/local/cuda
CUDA_HOME=/usr/local/cuda
CUDA_VISIBLE_DEVICES=0
CUDA_VISIBLE_DEVICES=0
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_xxx


```

</details>

### 🐛 Describe the bug

Long-text embeddings returned by vLLM are not unit-normalized when chunked processing is enabled, even when the request explicitly uses `use_activation=true`.

This affects more than the reported vector norm. It changes dot-product scores, threshold decisions, and comparisons between short and long inputs.

Using embeddings returned by the API, an exact long-text self-match produces a dot product of only `0.61587080`. The same vector has cosine similarity `1.0`, and its dot product becomes `1.0` after L2 normalization.

For example, a dot-product threshold of `0.8` rejects an exact match:

```text
same 128-token text used as query and document

raw vLLM dot product:       0.61587080  -> rejected
dot after L2 normalization: 1.00000000  -> accepted
threshold:                  0.8
```

All vectors in this comparison come directly from `POST /v1/embeddings`.

## Environment

- vLLM: `0.29.0`
- Model: `Qwen/Qwen3-Embedding-0.6B`
- GPU: NVIDIA L40S
- Endpoint: `POST /v1/embeddings`
- Runner: pooling

Pooler configuration:

```json
{
  "enable_chunked_processing": true,
  "max_embed_len": 512,
  "use_activation": true
}
```

## Reproduction

Start the server with `max_model_len=64` so a 128-token input is split into
two chunks:

```bash
MODEL_PATH=/path/to/Qwen3-Embedding-0.6B

CUDA_VISIBLE_DEVICES=0 \
vllm serve "$MODEL_PATH" \
  --served-model-name chunk-embed \
  --runner pooling \
  --convert embed \
  --host 127.0.0.1 \
  --port 8134 \
  --max-model-len 64 \
  --pooler-config \
    '{"enable_chunked_processing":true,"max_embed_len":512,"use_activation":true}' \
  --gpu-memory-utilization 0.20 \
  --max-num-seqs 8 \
  --enforce-eager
```

Use the script: [probe_vllm_chunked_embedding_normalization.py](https://github.com/user-attachments/files/32462928/probe_vllm_chunked_embedding_normalization.py)
Run:

```bash
python scripts/probe_vllm_chunked_embedding_normalization.py \
  --base-url http://127.0.0.1:8134 \
  --tokenizer "$MODEL_PATH" \
  --out /tmp/chunked_embedding_result.json
```

The reproducer builds two different 64-token sequences:

- `A`: repetitions of
  `apple orchard harvest fruit trees sunshine agriculture`;
- `B`: repetitions of
  `quantum mechanics particle wavefunction entanglement laboratory`.

It requests embeddings for:

1. `A` and `B` as ordinary 64-token inputs;
2. `A + B` as one 128-token long input;
3. the same returned vectors under representative dot-product scoring and
   threshold rules.

Every embedding request uses:

```json
{
  "model": "chunk-embed",
  "encoding_format": "float",
  "add_special_tokens": false,
  "use_activation": true
}
```

### 1. Exact long-text match fails a dot-product threshold

```text
query:    embedding(A + B)
document: embedding(A + B)

raw dot product:              0.61587080
dot after L2 normalization:   1.00000000
cosine similarity:            1.00000000
threshold:                    0.8
raw decision:                 reject
normalized decision:          accept
```

Every threshold in `(0.61587080, 1.0]` changes this decision.

### 2. A short query can fail against a long document containing it

```text
query:    embedding(A)
document: embedding(A + B)

raw dot product:              0.61584421
dot after L2 normalization:   0.78474047
threshold:                    0.7
raw decision:                 reject
normalized decision:          accept
```

Every threshold in `(0.61584421, 0.78474047]` changes this decision.

### 3. Exact-match ranking margin nearly disappears

Using `embedding(A + B)` as the query:

| Candidate | Raw dot product | Dot after normalization |
|---|---:|---:|
| Exact long document `A + B` | `0.61587080` | `1.00000000` |
| Short half-document `A` | `0.61584421` | `0.78474047` |

The exact-match margin is only `0.00002658` with the raw vLLM outputs. After
normalization it is `0.21525953`.

### 4. Long and short outputs have different norms

```text
ordinary A norm:          1.00000003
ordinary B norm:          1.00000004
chunked A+B norm:         0.78477436
explicitly normalized:    1.00000000
```

A magnitude threshold of `0.9` therefore accepts the ordinary embeddings but
rejects the chunked long embedding.

## Expected result

With `use_activation=true`, both ordinary and chunked embedding outputs should
be L2-normalized. The exact long-text self dot product should therefore be
approximately `1.0`.

The final aggregation should be equivalent to:

```python
final_embedding = F.normalize(
    weighted_sum / total_weight,
    p=2,
    dim=-1,
)
```

## Root cause

Each chunk passes through the normal embedding pooler and is L2-normalized
before aggregation.

`EmbedIOProcessor._post_process_chunked()` then computes a token-count-weighted
mean:

```python
final_embedding = weighted_sum / total_weight
```

That mean is returned directly. There is no final activation or
`F.normalize()` call after the chunks have been combined.

The mean of different unit vectors generally has a norm below 1. In this
reproduction, a dot product involving one affected long vector is scaled by
approximately `0.7848`; a long-vector self dot product is scaled by its
square, approximately `0.6159`.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by Ningke-Li** [comment]
I also compared the chunked output with the embeddings returned for its two individual chunks.

The 128-token input is the concatenation of two different 64-token sequences, `A` and `B`. With `use_activation=true`, the ordinary requests return unit vectors:

```text
norm(embedding(A)): 1.0000000303
norm(embedding(B)): 1.0000000355
```

The same tokens sent as one long `A + B` request are processed through the chunked path and return:

```text
norm(embedding(A + B)): 0.7847743598
```

The long output closely matches the token-count-weighted mean of the two ordinary chunk embeddings:

```text
norm(weighted_mean(A, B)):               0.7844531468
absolute norm difference from API output: 0.0003212130
maximum elementwise difference:           0.0020161755
```

The small difference is consistent with the chunks being evaluated in a different BF16 batch shape. Explicitly normalizing the aggregate restores the expected unit norm:

```text
norm(L2_normalize(weighted_mean(A, B))): 1.0
```

As a causal control, normalizing the returned long vector changes its magnitude but not its direction:

```text
cosine(raw_long, L2_normalize(raw_long)): 1.0000000000000002
raw self dot:                              0.6158707958
normalized self dot:                       1.0
```

This matches the implementation order:

1. each chunk passes through the embedding pooler and is normalized;
2. `EmbedIOProcessor._post_process_chunked()` computes `weighted_sum / total_weight`;
3. that aggregate is returned without another activation or `torch.nn.functional.normalize()` call.

Therefore the changed dot-product scores are explained specifically by the missing post-aggregation L2 normalization, rather than by a change in embedding direction.

**Comment by Ningke-Li** [comment]
#57498 seems to fix this issue. Will have a test.
